Refactor optimization post process

guilpier-code opened the pull request:
The main idea is to customize the post processes to be run, in regard to the weekly optimization to be run : the adequacy patch weekly optimization is not followed by the same post processes than the default weekly optimization. 

To do that : 
- **post process commands** : each post process used to be performed by a free function. This call was moved in the **execute()** methods of a command design pattern. These commands are meant to be added to a vector of post-process commands.
- **Post-process list (vector)** : depending of the kind of weekly optimization (adequacy path or default), we create a vector of post-process. This list is not supposed to be the same in either cases.
- **Running the post processes** : running the post processes is done in one line inside a simulation year.

guilpier-code committed:
move calculation of LMR violations from CSR source file to adq patch weekly problem

guilpier-code committed:
Merge remote-tracking branch 'remotes/origin/feature/adq-patch-final' into fix/move-lmr-viol-from-csr-to-adq-weekly-pb

guilpier-code committed:
Refactor Optimization post processes : first work : add abstractions for post-process commands and for a list of them.

guilpier-code committed:
Merge remote-tracking branch 'remotes/origin/feature/adq-patch-final' into fix/move-lmr-viol-from-csr-to-adq-weekly-pb

guilpier-code committed:
Refactor Optimization post processes : move the first post process in the list of post processes to be run

flomnes commented on src/solver/optimisation/post_process_commands.cpp:
```diff
@@ -0,0 +1,28 @@
+
+void DispatchableMarginPostProcessCmd::acquireOptRuntimeData(const struct optRuntimeData& opt_runtime_data)
+{
+    hourInYear_ = opt_runtime_data.year;
```
Is it a year or an hour ?

guilpier-code replied:
Exactly the problem of regression that I had and fixed !
Good eyes !
So this is corrected now

guilpier-code committed:
Refactor Optimization post processes : fixing a regression

guilpier-code committed:
Refactor Optimization post processes : moving the code of the post process associated to dispatchable margin into the command class

sylvlecl commented on src/solver/simulation/base_post_process.h:
```diff
@@ -0,0 +1,69 @@
+    unsigned int hourInTheYear = 0;
+};
+
+class basePostProcessCommand
```
For namings I would suggest to:
 - use consistently `UpperCamelCase` for classes/interfaces
 - remove the words `base` or `interface` from the root of hierarchies: it only hurts readability with longer names without real added value in the context of their use (we don't need to know it's an interface when we use it)
 - In this particular case, I would suggest to remove "command" word from the name, which is most often used for command line: maybe just `PostProcessor` ?

guilpier-code committed:
Merge remote-tracking branch 'remotes/origin/feature/adq-patch-final' into fix/reactor-opt-post-process

guilpier-code commented on src/solver/cmake/simulation.cmake:
```diff
@@ -34,13 +34,14 @@ set(SRC_SIMULATION
 		simulation/adequacy.cpp
 		simulation/economy.h
 		simulation/economy.cpp
+		simulation/base_post_process.h
```
Rename **base_post_process.h/.cpp** into **post_process.h/.cpp** (according to remark of @sylvlecl )

guilpier-code commented on src/solver/optimisation/post_process_commands.h:
```diff
@@ -0,0 +1,23 @@
+                                     unsigned int thread_number,
+                                     AreaList& areas);
+    void acquireOptRuntimeData(const struct optRuntimeData& opt_runtime_data) override;
+    void run() override;
```
Rename **run()** into **execute()**, to stick with the command design pattern, so that it can be more recognizable

guilpier-code replied:
Done

guilpier-code commented on src/solver/optimisation/optim_post_process_list.h:
```diff
@@ -0,0 +1,22 @@
+namespace Antares::Solver::Simulation
+{
+
+class OptPostProcessList : public interfacePostProcessList
```
Rename **OptPostProcessList** into **DefaultOptPostProcessList**.
Rename file **optim_post_process_list.h** into **default_opt_post_process_list.h**

guilpier-code committed:
Refactor Optimization post processes : some cleaning (see details in the commit's body)

- Narrow arguments of optimization post process functions. For example, study is passes while we only need the study's collection of areas
- Not pass the constant "nbHoursInAWeek" as an argument, rather make it a constant visible by every function using it.
- Remove some useless headers

guilpier-code committed:
Refactor Optimization post processes : fixing a crash

guilpier-code committed:
Refactor Optimization post processes : adding hydro levels update to list of post process

guilpier-code committed:
Refactor Optimization post processes : adding hydro remix to the list of post process

guilpier-code committed:
Refactor Optimization post processes : adding DTG margin for adq patch to the list of post process

guilpier-code committed:
Refactor Optimization post processes : forgot to remove adqPatchPostProcess function from curtailment sharing problem

guilpier-code committed:
Refactor Optimization post processes : adding hydro levels update (after remix) to list of post process

flomnes commented on src/solver/optimisation/adq_patch_post_process_list.h:
```diff
@@ -0,0 +1,23 @@
+                            SimplexOptimization splxOptimization,
+                            Calendar& calendar);
+
+    ~AdqPatchPostProcessList() = default;
```
Make this function `virtual`

guilpier-code replied:
Done : I made the destructor **virtual**

flomnes commented on src/solver/simulation/base_post_process.cpp:
```diff
@@ -0,0 +1,48 @@
+
+void interfacePostProcessList::runAll(const struct optRuntimeData& opt_runtime_data)
+{
+    for (auto it = post_process_list.begin(); it != post_process_list.end(); it++)
```
```suggestion
    for (auto command : post_process_list)
```

guilpier-code replied:
Done

flomnes commented on src/solver/simulation/base_post_process.cpp:
```diff
@@ -0,0 +1,48 @@
+{
+    for (auto it = post_process_list.begin(); it != post_process_list.end(); it++)
+    {
+        (*it)->acquireOptRuntimeData(opt_runtime_data);
+        (*it)->run();
```
```suggestion
        cmd->run(context);
```

guilpier-code replied:
You're right, better that way.
Done

flomnes reviewed (changes requested):
IMO there is no need for function `acquireOptRuntimeData`. Just call `run` with whatever context is necessary.

guilpier-code committed:
Refactor Optimization post processes : adding post process "interpolate water value" to list of post process

flomnes commented on src/solver/optimisation/optim_post_process_list.cpp:
```diff
@@ -0,0 +1,44 @@
+
+    : interfacePostProcessList(problemeHebdo, thread_number)
+{
+    post_process_list.push_back(std::make_unique<DispatchableMarginPostProcessCmd>(
+        problemeHebdo_, 
+        thread_number_, 
+        areas));
+    post_process_list.push_back(std::make_unique<HydroLevelsUpdatePostProcessCmd>(
+        problemeHebdo_, 
+        areas, 
+        false, 
+        false));
+    post_process_list.push_back(std::make_unique<RemixHydroPostProcessCmd>(
+        problemeHebdo_,
+        areas,
+        sheddingPolicy,
+        splxOptimization,
+        thread_number));
+    post_process_list.push_back(std::make_unique<HydroLevelsUpdatePostProcessCmd>(
+        problemeHebdo_,
+        areas,
+        true,
+        false));
+    post_process_list.push_back(std::make_unique<InterpolateWaterValuePostProcessCmd>(
+        problemeHebdo_,
+        areas,
+        calendar));
```
I know you don't like if..else much, but IMO it would be clearer to have
```cpp
post_process_list.push_back(...);
if (adqPatch)
    post_process_list.push_back(std::make_unique<DTGmarginForAdqPatchPostProcessCmd>(...);
post_process_list.push_back(...);
```
It would avoid code duplication.

flomnes replied:
You could hide this in a factory function
```cpp
postProcessListFactory(bool adqPatch) {
// ...
}
```

guilpier-code replied:
This should be discussed more deeply. 
Before taking a decision, let's look at **adequacy.cpp**, which has not been touched yet.

flomnes commented on src/solver/simulation/common-eco-adq.h:
```diff
@@ -88,16 +88,16 @@ void ComputeFlowQuad(Data::Study& study,
                            uint numSpace,
-                           uint hourInYear,
-                           uint nbHours);
+                           uint hourInYear);
```
This is context info

guilpier-code replied:
could you give some more explanations about what you mean ?

flomnes replied:
It should be picked up from `optRuntimeData`

guilpier-code replied:
This free function **RemixHydroForAllAreas(...)** is called from a post process command, which retrieves  **hourInYear** from **optRuntimeData**.
We could move the body of this function to the **run(...)** method of this post process command. 
But it would have some consequences, as  **RemixHydroForAllAreas(...)** is also called in **adequacy.cpp**. 
So I preferred postponing this refactoring.

flomnes commented:
undefined reference to `Antares::Solver::Simulation::RemixHydroForAllAreas(Antares::Data::AreaList const&, PROBLEME_HEBDO&, Antares::Data::SheddingPolicy, Antares::Data::SimplexOptimization, unsigned int, unsigned int)`

guilpier-code commented:
> undefined reference to `Antares::Solver::Simulation::RemixHydroForAllAreas(Antares::Data::AreaList const&, PROBLEME_HEBDO&, Antares::Data::SheddingPolicy, Antares::Data::SimplexOptimization, unsigned int, unsigned int)`

Yes, only on Ubuntu, **antares-solver** is built, but **tests-simple-study** fails to compile

guilpier-code committed:
Refactor Optimization post processes : adding post process "Hydro Levels Final Update" to list of post process

guilpier-code committed:
[skip ci] Refactor Optimization post processes : adding an empty CSR command to list of post process (to be filled later)

guilpier-code committed:
Refactor Optimization post processes : correction after code review

guilpier-code committed:
[skip ci] Refactor Optimization post processes : make destructors virtual
